### ISSUE TITLE
Discussion of IBEX memory usage to find way to reduce it

### ISSUE THREAD

**Issue by ChrisM-S** [ISISComputingGroup/IBEX#3574]
For doing a broadly similar job to SECI, the early IBEX systems are memory hungry.  This is not entirely surprising given that development priority has been given to conversion of machines and gaining system functionality above resource usage.

As we are starting to bump into memory resource issues across existing hardware (originally sized not for instrument control but for neutron data acquisition and storage needs) it is opportune to look at this now.

1) See if there are some quick gains can be made easily across the board.
2) Create some tickets for sensible modifications to our system which might be a little harder but are sort of obvious.  For example, allocating IOC memory on need/startup and dynamically rather than statically in 100s of processes as with the procserv system.

Acceptance Criteria:

1. Analysis of memory usage along the lines of people but with Java separated by actual process (this can probably done on a dev machine if it looks the same)
1. Discuss to see what the easiest way to reduce memory is without effecting performance.
1. Tickets have been created to reduce memory performance.

**Comment by ChrisM-S** [comment]
As a bit of supplementary information, I've pasted in a couple of "perfmon" plots given a comparison of the situation on two instruments of similar complexity one IBEX and one SECI.

### Memory usage for POLREF (on SECI still)
![process memory polref on seci](https://user-images.githubusercontent.com/3035405/45815684-4ebd0f80-bcd1-11e8-9560-64d6434cd7f3.png)
### Memory usage for POLARIS (on IBEX)
![process memory polaris on seci](https://user-images.githubusercontent.com/3035405/45815901-fa665f80-bcd1-11e8-85f3-ec7610ba4cbb.png)

**Comment by ChrisM-S** [comment]
@Tom-Willemsen I don't think there is an easy way to identify process command lines from the `process#n` given by the VMI performance counters (except possibly that the order of numbering follows process id value and time - increasing from the left on the plots above).

To get similar information in text form (including command line and process id), the WMI process information can be gathered as below.  Note that `PrivatePageCount` is apparently NOT a page count but a byte count, so multiplying by default page size (4096) just confuses things!

To get all process info:
`PS C:\Users\cms42> $PG =(Get-WmiObject -Credential $cred -Class win32_process -ComputerName ndxpolaris)`

To sum up processes (selecting  by name/size if required)
`((($PG | sort -Property privatepagecount -Descending | ?{($_.privatepagecount -gt 100000) -and ($_.name -like "Procserv.exe")} | Measure-Object -Property privatepagecount -sum).sum))/1MB`
or to print command line or anything else.
`$PG[n].commandline`

**Comment by John-Holt-Tessella** [comment]
We also need an estimate on the total memory use by IBEX when installed for the next 5 years so that setup machines can be bought with the right sized memory; otherwise, we will have to change the deploy for setup machine which current have 4Gb. This must include the effects of moving to Windows 10 on the setup machines.

**Comment by davidkeymer** [comment]
Reviewed by checking that the following tickets have been created as a result of the meeting:

- Control system behaviour with reducing amounts of available memory: https://github.com/ISISComputingGroup/IBEX/issues/3864
- Creating ProcServ processes only when required: https://github.com/ISISComputingGroup/IBEX/issues/3858
- Capping memory used by Java processes: https://github.com/ISISComputingGroup/IBEX/issues/3859
- Increase RAM of control machines: https://github.com/ISISComputingGroup/IBEX/issues/3860
- Increase RAM of setup machines: https://github.com/ISISComputingGroup/IBEX/issues/3861
- Create system test to monitor overall memory usage: https://github.com/ISISComputingGroup/IBEX/issues/3862
